File: LC089GrayCode.java
package leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode.com/problems/gray-code/
 */
public class LC089GrayCode {
    public List<Integer> grayCode(int bits) {
        if (bits == 0) {
            ArrayList<Integer> list = new ArrayList<Integer>();
            list.add(0);
            return list;
        }
        ArrayList<Integer> list = new ArrayList<>();
        list.add(0);
        list.add(1);
        int shifted = 2;
        while (bits > 1) {
            bits--;
            int n = list.size();
            for (int i = n - 1; i >= 0; i--) {
                int value = list.get(i);
                list.add(shifted | value);
            }
            shifted <<= 1;
        }
        return list;
    }

    public static void main(String[] arguments) {
        LC089GrayCode solution = new LC089GrayCode();
        System.out.println(solution.grayCode(4));
    }
}

File: algorithms/code/leetcode/lc089_gray_code/lc089_gray_code.py
#!/usr/bin/env python3
# https://leetcode.com/problems/gray-code/

import unittest
from typing import List


class Solution:
    def grayCode(self, bits: int) -> List[int]:
        if bits == 0:
            listed = [0]
            return listed
        listed = [0, 1]
        shifted = 2
        while bits > 1:
            bits -= 1
            n = len(listed)
            for i in range(n - 1, -1, -1):
                value = listed[i]
                listed.append(shifted | value)
            shifted <<= 1
        return listed


class TestCode(unittest.TestCase):
    def test_4(self):
        expected = [0, 1, 3, 2, 6, 7, 5, 4, 12, 13, 15, 14, 10, 11, 9, 8]
        self.assertListEqual(expected, Solution().grayCode(4))

    def test_0(self):
        expected = [0]
        self.assertListEqual(expected, Solution().grayCode(0))
